addons/mrp/models/mrp_production.py

```diff
@@ -1713,12 +1714,22 @@ def _default_amounts(production):
             for workorder in production.workorder_ids:
                 initial_workorder_remaining_qty.append(max(workorder.qty_produced - workorder.qty_production, 0))
                 workorder.qty_produced = min(workorder.qty_produced, workorder.qty_production)
-            workorders_len = len(bo.workorder_ids)
+            workorders_len = len(production.workorder_ids)
             for index, workorder in enumerate(bo.workorder_ids):
-                remaining_qty = initial_workorder_remaining_qty[index // workorders_len]
+                remaining_qty = initial_workorder_remaining_qty[index % workorders_len]
                 if remaining_qty:
                     workorder.qty_produced = max(workorder.qty_production, remaining_qty)
                     initial_workorder_remaining_qty[index % workorders_len] = max(remaining_qty - workorder.qty_produced, 0)
+                    if not workorder.qty_remaining:
+                        workorders_to_cancel += workorder
+                    elif workorders_to_update[-1:].production_id != workorder.production_id:
+                        workorders_to_update += workorder
+                elif workorder.qty_remaining and workorders_to_update[-1:].production_id != workorder.production_id:
+                    workorders_to_update += workorder
+
+        workorders_to_cancel.action_cancel()
+        for workorder in workorders_to_update:
```

Comment:
For performance reason, I would set the state as follows:
```python
workorders.filtered(lambda wo: wo.next_work_order_id.production_availability == 'assigned').state = 'ready' 
workorders.filtered(lambda wo: wo.next_work_order_id.production_availability != 'assigned').state = 'waiting'
```

Reply:
good point 
done :)